test/docker/presync-webhook-server/oci-signature-verification-server.go

```diff
@@ -0,0 +1,154 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+)
+
+const (
+	imageToSync   = "configsync.gke.io/image-to-sync"
+	publicKeyPath = "/cosign-key/cosign.pub"
+)
+
+// getAnnotations extracts annotations from the raw JSON data.
+func getAnnotations(raw []byte) (map[string]string, error) {
+	var metadata map[string]interface{}
+	if err := json.Unmarshal(raw, &metadata); err != nil {
+		return nil, err
+	}
+	if annotations, ok := metadata["metadata"].(map[string]interface{})["annotations"].(map[string]interface{}); ok {
+		annotationsMap := make(map[string]string)
+		for k, v := range annotations {
+			annotationsMap[k] = fmt.Sprintf("%v", v)
+		}
+		return annotationsMap, nil
+	}
+	klog.Infof("no annotations found in the objects %s", metadata["name"])
+	return nil, nil
+}
+
+func verifyImageSignature(ctx context.Context, image string) error {
+	if image == "" {
+		return nil
+	}
+
+	pubKey, err := signature.LoadPublicKey(ctx, publicKeyPath)
+	if err != nil {
+		return fmt.Errorf("error loading public key: %v", err)
+	}
+
+	googleAuth, err := google.NewEnvAuthenticator(ctx)
+	if err != nil {
+		return err
+	}
+
+	opts := &cosign.CheckOpts{
+		RegistryClientOpts: []ociremote.Option{ociremote.WithRemoteOptions(remote.WithAuth(googleAuth))},
+		SigVerifier:        pubKey,
+		IgnoreTlog:         true,
+	}
+
+	ref, err := name.ParseReference(image)
+	if err != nil {
+		return fmt.Errorf("failed to parse image reference: %v", err)
+	}
+	_, _, err = cosign.VerifyImageSignatures(ctx, ref, opts)
+	if err != nil {
+		return fmt.Errorf("image verification failed for %s: %v", image, err)
+	}
+
+	klog.Infof("Image %s verified successfully", image)
+	return nil
+}
+
+func handleWebhook(w http.ResponseWriter, r *http.Request) {
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		klog.Errorf("Failed to read request body: %v", err)
+		http.Error(w, "Failed to read request body", http.StatusBadRequest)
+		return
+	}
+
+	var admissionReview admissionv1.AdmissionReview
+	if err := json.Unmarshal(body, &admissionReview); err != nil {
+		klog.Errorf("Failed to unmarshal admission review: %v", err)
+		http.Error(w, "Failed to unmarshal admission review", http.StatusBadRequest)
+		return
+	}
+
+	response := &admissionv1.AdmissionResponse{
+		UID: admissionReview.Request.UID,
+	}
+
+	oldAnnotations, err := getAnnotations(admissionReview.Request.OldObject.Raw)
+	if err != nil {
+		klog.Errorf("Failed to extract old annotations: %v", err)
+		response.Result = &metav1.Status{
+			Message: fmt.Sprintf("Failed to extract old annotations: %v", err),
+		}
+		response.Allowed = false
+	}
+
+	newAnnotations, err := getAnnotations(admissionReview.Request.Object.Raw)
+	if err != nil {
+		klog.Errorf("Failed to extract new annotations: %v", err)
+		response.Result = &metav1.Status{
+			Message: fmt.Sprintf("Failed to extract new annotations: %v", err),
+		}
+		response.Allowed = false
+	}
+
+	if newAnnotations[imageToSync] != oldAnnotations[imageToSync] {
```

Comment:
This may lead to a nil pointer if `getAnnotations` returns nil.

We can update `getAnnotations` to `getAnnotation` by the annotation key.

Reply:
Done